x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/SecurityExtension.java

```diff
@@ -46,8 +66,10 @@
      *
      * Only one installed extension may have an authentication failure handler. If more than
      * one extension returns a non-null handler, an error is raised.
+     *
+     * @param components Access to components that may be used to build the handler
      */
-    default AuthenticationFailureHandler getAuthenticationFailureHandler() {
+    default AuthenticationFailureHandler getAuthenticationFailureHandler(SecurityComponents components) {
         return null;
     }
```

Comment:
I think we've got a chance to avoid breaking the java interface in a minor, although we don't commit to it in general.
I would think of adding a new Service provider getter with the same name but different parameters, deprecate the existing one, and have a static method that multiplexes between the two, for example:

```
@Deprecated  
default AuthenticationFailureHandler getAuthenticationFailureHandler() {
  return null;
}

default AuthenticationFailureHandler getAuthenticationFailureHandler(SecurityComponents components) {
  return null;
}

static AuthenticationFailureHandler getAuthenticationFailureHandler(SecurityExtension securityExtension, SecurityComponents components) {
  AuthenticationFailureHandler authnHandler = securityExtension.getAuthenticationFailureHandler();
  If (authnHandler != null) {
    if (securityExtension.getAuthenticationFailureHandler(components) != null) {
      throw new IllegalArgumentException("Cannot implement both handlers");
    }
    // log deprecation
    return authnHandler;
  }
  return securityExtension.getAuthenticationFailureHandler(components);
}
```

What do you think, is it worth the complexity?

Reply:
I don't think it's worth it - plugins need to be recompiled anyway, and this will force plugin authors to acknowledge the change (assuming they have an `@Override` annotation).

We _could_ do it, but this is a case where we'd be carrying a heavy burden to save a few people from making a very minor change.

Reply:
Okay, sounds fair.